File: Retaguarda/Torneio.Domain/Entities/DoacaoPatrocinador.cs
using Torneio.Domain.Enums;

namespace Torneio.Domain.Entities;

public class DoacaoPatrocinador
{
    public Guid Id { get; private set; }
    public Guid TorneioId { get; private set; }
    public Guid? PatrocinadorId { get; private set; }
    public string NomePatrocinador { get; private set; } = null!;
    public TipoDoacaoPatrocinador Tipo { get; private set; }
    public string Descricao { get; private set; } = null!;
    public decimal? Quantidade { get; private set; }
    public decimal? Valor { get; private set; }
    public string? Observacao { get; private set; }
    public DateTime DataDoacao { get; private set; }
    public DateTime CriadoEm { get; private set; }

    private DoacaoPatrocinador() { }

    public static DoacaoPatrocinador Criar(
        Guid torneioId,
        Guid? patrocinadorId,
        string nomePatrocinador,
        TipoDoacaoPatrocinador tipo,
        string descricao,
        decimal? quantidade,
        decimal? valor,
        string? observacao,
        DateTime dataDoacao)
    {
        var entidade = new DoacaoPatrocinador
        {
            Id = Guid.NewGuid(),
            TorneioId = torneioId,
            CriadoEm = DateTime.UtcNow
        };
        entidade.Atualizar(patrocinadorId, nomePatrocinador, tipo, descricao, quantidade, valor, observacao, dataDoacao);
        return entidade;
    }

    public void Atualizar(
        Guid? patrocinadorId,
        string nomePatrocinador,
        TipoDoacaoPatrocinador tipo,
        string descricao,
        decimal? quantidade,
        decimal? valor,
        string? observacao,
        DateTime dataDoacao)
    {
        PatrocinadorId = patrocinadorId;
        NomePatrocinador = nomePatrocinador.Trim();
        Tipo = tipo;
        Descricao = descricao.Trim();
        Quantidade = quantidade;
        Valor = valor;
        Observacao = string.IsNullOrWhiteSpace(observacao) ? null : observacao.Trim();
        DataDoacao = dataDoacao.Date;
    }
}

File: App/torneio_app/lib/core/models/doacao_patrocinador.dart
class DoacaoPatrocinador {
  final String id;
  final String torneioId;
  final String? patrocinadorId;
  final String nomePatrocinador;
  final String tipo;
  final String descricao;
  final double? quantidade;
  final double? valor;
  final String? observacao;
  final DateTime dataDoacao;

  const DoacaoPatrocinador({
    required this.id,
    required this.torneioId,
    this.patrocinadorId,
    required this.nomePatrocinador,
    required this.tipo,
    required this.descricao,
    this.quantidade,
    this.valor,
    this.observacao,
    required this.dataDoacao,
  });

  bool get geraReceita => tipo.toLowerCase() == 'dinheiro' && (valor ?? 0) > 0;

  factory DoacaoPatrocinador.fromJson(Map<String, dynamic> json) => DoacaoPatrocinador(
        id: json['id'] as String,
        torneioId: json['torneioId'] as String,
        patrocinadorId: json['patrocinadorId'] as String?,
        nomePatrocinador: json['nomePatrocinador'] as String? ?? '',
        tipo: json['tipo'] as String? ?? 'Dinheiro',
        descricao: json['descricao'] as String? ?? '',
        quantidade: (json['quantidade'] as num?)?.toDouble(),
        valor: (json['valor'] as num?)?.toDouble(),
        observacao: json['observacao'] as String?,
        dataDoacao: DateTime.tryParse(json['dataDoacao'] as String? ?? '') ?? DateTime.now(),
      );
}
